Mathlib/MeasureTheory/Function/UnifTight.lean

```diff
@@ -0,0 +1,511 @@
+    {f : α → β} (hf : Memℒp f p μ)
+    {ε : ℝ} (hε : 0 < ε) :
+    ∃ (s : Set α) (_ : MeasurableSet s) (_ : μ s < ∞),
+      snorm (sᶜ.indicator f) p μ ≤ ENNReal.ofReal ε := by
+  -- The proof unwraps `Memℒp f p μ` and applies the analogous result for `lintegral`.
+  -- do some arithmetic that will come in useful
+  have hp_pos := zero_lt_one.trans_le hp_one
+  have hp_nz := hp_pos.ne'
+  have hrp_pos : 0 < p.toReal := ENNReal.toReal_pos hp_nz hp_top
+  have hirp_pos : 0 < 1 / p.toReal := div_pos (by norm_num) hrp_pos
+  have hεp : 0 < ε ^ p.toReal := by simp only [Real.rpow_pos_of_pos, hε]
+  -- decode Memℒp into a.e. strong measurability and finite snorm
+  obtain ⟨haesmf, hsnf⟩ := hf
+  -- transform snorm to lintegral
+  rw [snorm_eq_lintegral_rpow_nnnorm hp_nz hp_top] at hsnf
+  replace hsnf := (rpow_lt_top_iff_of_pos hirp_pos).mp hsnf
+  -- get a.e. measurability for the integrand
+  -- XXX: Why does `AEStronglyMeasurable.ennnorm` only give the weaker AEMeasurable?
+  --      It would make sense to me to use `haesmf.ennnorm.aemeasurable` below.
+  have haemnf := haesmf.ennnorm
+  have haemnpf := haemnf.pow_const p.toReal
+  -- use core result for lintegral (needs only AEMeasurable), the target estimate will be in `hsfε`
+  obtain ⟨s, hms, hμs, hsfε⟩ := lintegral_indicator_compl_le haemnpf hsnf hεp
+  rw [← lintegral_indicator _ hms.compl] at hsfε
+  use s, hms, hμs
+  -- move indicator through function compositions, XXX: is this simp-able?
+  rw [← Function.comp_def (fun x : ℝ≥0∞ => x ^ p.toReal)] at hsfε
+  rw [← Function.comp_def ENNReal.ofNNReal] at hsfε
+  rw [← Function.comp_def nnnorm] at hsfε
+  rw [sᶜ.indicator_comp_of_zero (@ENNReal.zero_rpow_of_pos p.toReal hrp_pos)] at hsfε
+  rw [sᶜ.indicator_comp_of_zero ENNReal.coe_zero] at hsfε
+  rw [sᶜ.indicator_comp_of_zero nnnorm_zero] at hsfε
+  rw [Function.comp_def nnnorm] at hsfε
+  rw [Function.comp_def ENNReal.ofNNReal] at hsfε
+  rw [Function.comp_def (fun x : ℝ≥0∞ => x ^ p.toReal)] at hsfε
+  -- commute ENNReal coersion with rpow, use rpow monotonicity
+  rw [← ofReal_rpow_of_pos (by assumption)] at hsfε
+  rw [← ENNReal.rpow_le_rpow_iff hirp_pos, ENNReal.rpow_inv_rpow_self hrp_pos.ne'] at hsfε
+  -- convert lintegral to snorm
+  rw [← snorm_eq_lintegral_rpow_nnnorm hp_nz hp_top] at hsfε
+  exact hsfε
+
+/-- A constant function is tight. -/
+theorem unifTight_const {g : α → β} (hp : 1 ≤ p) (hp_ne_top : p ≠ ∞) (hg : Memℒp g p μ) :
+    UnifTight (fun _ : ι => g) p μ := by
+  intro ε hε
+  by_cases hε_top : ε = ∞
+  · exact ⟨∅, (by measurability), fun _ => hε_top.symm ▸ le_top⟩
+  have hrε : 0 < ε.toReal := ENNReal.toReal_pos hε.ne' hε_top
+  obtain ⟨s, _, hμs, hgε⟩ := hg.snorm_indicator_compl_le hp hp_ne_top hrε
+  exact ⟨s, ne_of_lt hμs, fun _ => hgε.trans_eq (ENNReal.ofReal_toReal hε_top)⟩
+
+/-- A single function is tight. -/
+theorem unifTight_subsingleton [Subsingleton ι] (hp_one : 1 ≤ p) (hp_top : p ≠ ∞)
```

Comment:
The usual trick here is to add `unifTight_of_const` assuming `∀ i j, f i = f j` instead of dealing with `fun _ : ι ↦ g`. Then you prove the `Subsingleton` version by `unifTight_of_const fun i j ↦ congr_arg f (Subsingleton.elim i j)`.

Reply:
Probably this theorem can be gotten rid of altogether. It is only used in one place, `unifTight_fin`. Here I closely replicated the analogous code from `MeasureTheory.Function.UniformIntegrable`. If this code is to be updated, perhaps it would be better to do a common refactor of the `unifTight_*` and `unifIntegrable_*` theorems after this PR is accepted.